Refactor: Modularize code and apply best practices

johnqherman opened the pull request.

johnqherman committed:
Refactor: Modularize code and apply best practices

thmsrmbld commented on src/camera.py:
```diff
@@ -0,0 +1,152 @@
+class Camera:
+    """Camera class to handle camera-related operations."""
+
+    def __init__(self, url: str):
```
Perhaps a short docstring to express why the initialiser does what it does when the class boots, for extra context.

thmsrmbld commented on src/camera.py:
```diff
@@ -0,0 +1,152 @@
+        if self.stream_url is None:
+            return False
+
+        return all(
```
This return using the `all` evaluation is good -  pythonic.

thmsrmbld commented on src/camera.py:
```diff
@@ -0,0 +1,152 @@
+        )
+
+    def _get_camera_details(self) -> Optional[str]:
+        """Extracts the camera details."""
```
Which camera details?

thmsrmbld commented on src/tests/tests.py:
```diff
@@ -0,0 +1,63 @@
+@pytest.mark.parametrize(
+    "details, expected_output",
+    [
+        (
```
To make these bonkers strings more readable you could easily move them out into a test_constants.py file. Which will make your parametrization much simpler to read.

thmsrmbld commented on src/shared.py:
```diff
@@ -0,0 +1,30 @@
+def exponential_backoff(
+    attempts: int, initial_delay: int, factor: int, exception_types: Tuple[Type[Exception], ...]
+) -> Callable:
+    def exponential_backoff_decorator(func: Callable) -> Callable:
```
This is great. Decorators can feel complicated sometimes when you open them up - so I'd just recommend a docstring and a couple of inline comments to to explain why it's doing what it does.

Not too much, just a help for future you.

thmsrmbld commented on src/utils.py:
```diff
@@ -0,0 +1,107 @@
+    """Generates a tweet text based on the camera information and flag."""
+    city = camera_info["city"] if camera_info["city"] != "-" else "Unknown"
+    region = camera_info["region"] if camera_info["region"] != "-" else "Unknown"
+    country = (
```
This comprehension feels a bit odd to me because of all the .replace() calls. It looks like it probably works fine but I guess I'd expect to have a dedicated utility function that had a list of pre-defined str values that you knew you'd need to modify, that you could arbitrarily feed stuff into and get something out of.

Nothing wrong with it in practice really but to my mind it feels like it could be made a little more sensible. Any ideas?

johnqherman replied:
I went ahead and created a dedicated utility function called `replace_substrings` that takes a string and a dictionary of mappings as input, then replaces the substrings in the string based on the given mappings. I also replaced the comprehension in the create_tweet_text function to use said `replace_substrings` function.

```python
def replace_substrings(string: str, mappings: dict[str, str]) -> str:
    """Replaces substrings in a string based on a dictionary of mappings."""
    for old, new in mappings.items():
        string = string.replace(old, new)
    return string
```

thmsrmbld commented on src/utils.py:
```diff
@@ -0,0 +1,107 @@
+    else:
+        location = f"{city}, {country}"
+
+    if city == "Unknown" and region == "Unknown" and country == "United States":
```
One -liner comment to explain why these differ would be good.

thmsrmbld commented on src/main.py:
```diff
@@ -0,0 +1,49 @@
+        if tweet_posted_successfully:
+            logger.info("waiting for an hour...")
+            time.sleep(c.SLEEP_DURATION)
+        else:
```
On the `else` here at the end of the loop you may want a `logger` if it fails to explain what happened / the conditions of the failure.

thmsrmbld commented on src/constants.py:
```diff
@@ -0,0 +1,56 @@
+import os
```
You may want in-line type annotations for your constants in this file so you can see their data types at a glance.

thmsrmbld commented on src/camera.py:
```diff
@@ -0,0 +1,152 @@
+            return None
+
+        details = self.details
+
```
It's not completely clear to me what all these find() operations are doing here. What is the problem this block solves? I may be able to recommend a less repetitive method.

johnqherman replied:
The original code used the `find()` function repeatedly to figure out where each piece of info was in the string, then sliced the string to get what it needed.

I rewrote the function to now loop over a list of the info I want and use `find()` inside the loop to get the start and end positions for each piece. This way, I don't need to write the same code over and over again for each piece of info.

```python
    def _parse_camera_details(self) -> Dict[str, str]:
        """Parses the camera details and returns the camera info as a dictionary."""

        details = self.details if self.details is not None else None
        camera_info = {}

        if details is not None:
            for key in ["City", "Region", "Country", "Country code"]:
                start = details.find(f"{key}: ") + len(f"{key}: ")
                end = details.find(" ", start)
                value = details[start:end].strip()
                camera_info[key] = value

        return camera_info
```

johnqherman replied:
I've modified this further to correctly parse the details string by searching for the correct delimiters (newlines) and not relying on a specific order of keys. This ensures that the function the details string is parsed accurately regardless of the order of keys or the presence of spaces within values.

```python
    def _parse_camera_details(self) -> Optional[Dict[str, str]]:
        """Parses the camera details and returns the camera info as a dictionary."""
        details = self.details if self.details is not None else None
        camera_info = {}

        if details is not None:
            keys = [("City", "Latitude"), ("Region", "City"), ("Country", "Country code"), ("Country code", "Region")]
            for key, next_key in keys:
                start = details.find(f"{key}:") + len(f"{key}:")
                end = details.find(f"{next_key}:")
                value = details[start:end].strip()
                camera_info[key] = value
        return camera_info
```

thmsrmbld commented on src/camera.py:
```diff
@@ -0,0 +1,152 @@
+        factor=s.RETRY_DELAY_FACTOR,
+        exception_types=(SaveImageException, FetchCamerasException),
+    )
+    def _save_image(self, image_file_path: str, camera_url: str, request_headers: Dict[str, str]) -> bool:
```
Very impressed you got the back-off working here bud. (I haven't tested it but assume it does from the code...)

Did you find it particularly challenging?

johnqherman replied:
Thanks! Working on exponential backoff was a bit challenging, but not too tough. The main thing was getting a grip on decorators and how to use them correctly. Once I got that part down, putting together the backoff code was pretty smooth.

The decorator function wraps the original function and handles the backoff logic. If one of the specified exceptions pops up, it retries the function after waiting for the delay. The delay increases exponentially by multiplying it by the factor after each unsuccessful attempt. This keeps going until the function works or we hit the max attempts.

thmsrmbld commented on src/camera.py:
```diff
@@ -0,0 +1,152 @@
+        image = cv2.imread(image_file_path)
+
+        if image is None:
+            logging.error("image is empty. skipping...")
```
You probably want the file path in the log here if you have access to it so you know in your logs which was empty.

thmsrmbld commented on .pre-commit-config.yaml:
```diff
@@ -0,0 +1,32 @@
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+
+  - repo: https://github.com/pycqa/flake8
```
You may want to change flake8 to ruff - it's newer, faster and in rust:

```
      # python code linting
      - id: ruff
        entry: poetry run ruff --fix yunojuno
        language: system
        name: ruff      - Python code linting
        types: [python]
```

thmsrmbld commented on src/camera.py:
```diff
@@ -0,0 +1,152 @@
+        if saved_successfully:
+            if not self._image_is_solid_color(image_file_path=image_file_path):
+                return True
+            else:
```
This `else` on 148 is probably unnecessary due to your return on the previous line which terminates the flow anyway. You could move the below else code up and outdent it as the path for any False.

thmsrmbld commented on src/tests/tests.py:
```diff
@@ -0,0 +1,63 @@
+        ),
+    ],
+)
+def test_parse_camera_details(sample_class: Camera, details: str, expected_output: Dict[str, str]) -> None:
```
This is the right approach for the tests. There are a lot of other standalone functions in your app here that you could also quite easily unit test, which is worth doing.

johnqherman committed:
Refactor code for readability and optimization + enhance documentation

This commit implements several improvements to the existing codebase for better readability, optimization, and documentation. The changes include:

- Migrated from `flake8` to `ruff` for linting: `ruff` provides similar functionality but is newer and faster.

- Refactored the '_parse_camera_details' function for improved readability and maintainability. It now iterates over key pairs, reducing redundancy and making future adjustments easier.

- Introduced type hints to `constants.py` and `settings.py` for improved code clarity.

- Reorganized `settings.py` to match the order in which settings are used in the code. This should make it easier to find and change settings as needed.

- Improved the `exponential_backoff` function by adding docstrings for better understanding of its purpose, parameters, and return value. Also added inline comments for better clarity of the code execution flow.

- Significantly improved testing with better organization, moving tests into a class for cleaner separation. Hard-coded test parameters have been replaced with constant imports for better readability and maintainability. Additional test cases were also added to improve coverage, including testing the utility functions, with docstrings added to all tests to clarify their purpose.

- Extracted the logic for replacing substrings in the country string into a new function called `replace_substrings`, which takes a string and a dictionary of mappings and replaces all occurrences of the keys in the string with their corresponding values.

- Improved exception logging in `post_to_twitter` function to include the exception type for better troubleshooting.

johnqherman committed:
Moved country_replacements to constants.py
